In [8]:
# =========================
# 0️⃣ Imports
# =========================
import pandas as pd
import numpy as np
from datetime import datetime

# =========================
# 1️⃣ Load processed data
# =========================
data_path = "../phase2/processed_data/race_data_processed.csv"
df_clean = pd.read_csv(data_path, parse_dates=['Date', 'birthdate'])

print(f"Loaded data shape: {df_clean.shape}")
print(df_clean.head())

# =========================
# 2️⃣ Basic info & types
# =========================
print("\n=== Data Types & Info ===")
print(df_clean.info())

# =========================
# 3️⃣ Check missing values
# =========================
missing_pct = (df_clean.isnull().sum() / len(df_clean) * 100).sort_values(ascending=False)
missing_pct = missing_pct[missing_pct > 0]

print("\n=== Missing Values ===")
print(missing_pct if not missing_pct.empty else "✅ No missing values")

# =========================
# 4️⃣ Target variable distribution
# =========================
print("\n=== Target Variable Distribution ('is_top_10') ===")
print(df_clean['is_top_10'].value_counts())
print("Percentage:")
print(df_clean['is_top_10'].value_counts(normalize=True)*100)

# =========================
# 5️⃣ Numeric columns: ranges & extreme values
# =========================
numeric_cols = df_clean.select_dtypes(include=[np.number]).columns.tolist()
print("\n=== Numeric Columns Summary ===")
for col in numeric_cols:
    print(f"\nColumn: {col}")
    print(f"Min: {df_clean[col].min()}, Max: {df_clean[col].max()}, Mean: {df_clean[col].mean():.2f}, Std: {df_clean[col].std():.2f}")

# =========================
# 6️⃣ Date consistency
# =========================
print("\n=== Date Consistency ===")
print(f"Date range: {df_clean['Date'].min()} to {df_clean['Date'].max()}")
print(f"Missing dates: {df_clean['Date'].isna().sum()} rows")

# =========================
# 7️⃣ Check duplicates
# =========================
dup_rows = df_clean.duplicated(subset=['Rider', 'Race_ID', 'Stage_Number'], keep=False)
print(f"\nDuplicate race entries: {dup_rows.sum()}")
if dup_rows.any():
    display(df_clean[dup_rows].head())

# =========================
# 8️⃣ Age consistency
# =========================
age_diff = (df_clean['age_at_race'] - df_clean['Age']).abs()
print(f"\nMax difference between 'Age' and 'age_at_race': {age_diff.max()}")
print(f"Rows with difference > 1 year: {(age_diff > 1).sum()}")

# =========================
# 9️⃣ String columns check
# =========================
str_cols = df_clean.select_dtypes(include=[object]).columns.tolist()
for col in str_cols:
    sample_values = df_clean[col].dropna().astype(str).sample(5, random_state=1).tolist()
    print(f"\nColumn: {col}, sample values: {sample_values}")

# =========================
# 🔟 Rolling metrics check
# =========================
rolling_cols = ['rolling_avg_rank_5', 'rolling_avg_rank_10', 'recent_top10_count', 'recent_top10_rate']
print("\n=== Rolling Metrics Check ===")
for col in rolling_cols:
    if col in df_clean.columns:
        print(f"\nColumn: {col}")
        print(f"NaNs: {df_clean[col].isna().sum()}")
        print(f"Min: {df_clean[col].min()}, Max: {df_clean[col].max()}, Mean: {df_clean[col].mean():.2f}")

# =========================
# 1️⃣1️⃣ pps and rdr columns check
# =========================
pps_rdr_cols = [c for c in df_clean.columns if c.startswith('pps_') or c.startswith('rdr_')]
print("\n=== PPS & RDR Columns Check ===")
for col in pps_rdr_cols:
    non_numeric = df_clean[col].apply(lambda x: not isinstance(x, (int, float)))
    print(f"{col}: {non_numeric.sum()} non-numeric values")


Loaded data shape: (219423, 54)
  Timelag                 Rider  Age                           Team   Time  \
0   +0:10  AAEN JØRGENSEN Jonas   25  Team Saxo Bank - Tinkoff Bank   0:00   
1   +3:05  AAEN JØRGENSEN Jonas   25  Team Saxo Bank - Tinkoff Bank   3:55   
2   +3:13  AAEN JØRGENSEN Jonas   25  Team Saxo Bank - Tinkoff Bank   0:00   
3  +17:21  AAEN JØRGENSEN Jonas   25  Team Saxo Bank - Tinkoff Bank  14:16   
4  +22:50  AAEN JØRGENSEN Jonas   25  Team Saxo Bank - Tinkoff Bank   5:31   

                Race_Name                           Stage_Name       Date  \
0  Santos Tour Down Under             Stage 1 | Prospect-Clare 2012-01-17   
1  Santos Tour Down Under          Stage 2 | Lobethal-Stirling 2012-01-18   
2  Santos Tour Down Under        Stage 3 | Unley-Victor Harbor 2012-01-19   
3  Santos Tour Down Under            Stage 4 | Norwood-Tanunda 2012-01-20   
4  Santos Tour Down Under  Stage 5 | McLaren-Old Willunga Hill 2012-01-21   

  Stage_Type     Start  ... Pnt_max 

,Timelag,Rider,Age,Team,Time,Race_Name,Stage_Name,Date,Stage_Type,Start,...,Pnt_max,Pnt_std,Pnt_race_count,Pnt_points_per_race,age_at_race,race_tier,rolling_avg_rank_5,rolling_avg_rank_10,recent_top10_count,recent_top10_rate
143,+0:10,ACEVEDO Janier,28,Garmin Sharp,0:00,Volta Ciclista a Catalunya,Stage 1 | Calella-Calella,2014-03-24,RR,Calella,...,10.0,1.601008,2,10.0,28.0,3,28.6,33.0,0.0,0.0
144,+0:23,ACEVEDO Janier,28,Garmin Sharp,0:03,Volta Ciclista a Catalunya,Stage 2 | Mataró-Girona,2014-03-25,RR,Mataró,...,10.0,1.601008,2,10.0,28.0,3,30.6,33.5,0.0,0.0
145,+1:43,ACEVEDO Janier,28,Garmin Sharp,1:43,Volta Ciclista a Catalunya,Stage 3 | Banyoles-La Molina (Alp),2014-03-26,RR,Banyoles,...,10.0,1.601008,2,10.0,28.0,3,39.0,37.7,0.0,0.0
146,+19:12,ACEVEDO Janier,28,Garmin Sharp,17:33,Volta Ciclista a Catalunya,Stage 4 | Alp-Vallter 2000-Setcases,2014-03-27,RR,Alp,...,10.0,1.601008,2,10.0,28.0,3,45.6,40.4,0.0,0.0
147,+20:48,ACEVEDO Janier,28,Garmin Sharp,1:36,Volta Ciclista a Catalunya,Stage 5 | Llanars–Vall de Camprodon-Valls,2014-03-28,RR,Llanars–Vall de Camprodon,...,10.0,1.601008,2,10.0,28.0,3,64.8,50.2,0.0,0.0



Max difference between 'Age' and 'age_at_race': 9.095824777549623
Rows with difference > 1 year: 108906

Column: Timelag, sample values: ['+1:06:27', '+3:19:08', '+1:08', '+1:24:22', '+33:29']

Column: Rider, sample values: ['BUCHMANN Emanuel', 'BOUCHER David', 'MAES Nikolas', 'CIEŚLIK Paweł', 'AMADOR Andrey']

Column: Team, sample values: ['Team Giant - Alpecin', 'Etixx - Quick Step', 'Movistar Team', 'CCC Team', 'Team Europcar']

Column: Time, sample values: ['1:32', '0:05', '9:49', '0:00', '24:50']

Column: Race_Name, sample values: ['Tour de France', 'Vattenfall Cyclassics', 'Tour of Beijing', 'Volta Ciclista a Catalunya', 'Vuelta a España']

Column: Stage_Name, sample values: ['Stage 7 | Dubai Safari Park-City Walk', 'Stage 1 | Unley-Lyndoch', 'Stage 4 | Gansingen-Gstaad', 'Stage 4 | Olten-Gansingen', 'Stage 5 | Megève-Megève']

Column: Stage_Type, sample values: ['RR', 'RR', 'RR', 'RR', 'RR']

Column: Start, sample values: ['Pau', 'Hamburg', 'Chong Li', 'Girona', 'Belorado']

Co

Issues to Note

Age discrepancy

Max difference between Age and age_at_race = 9.1 years.

108,906 rows have >1 year difference.

Likely caused by missing/incorrect birthdate imputation.

Duplicate entries

Duplicates exist in (Rider, Race_ID, Stage_Number) — may need careful deduplication depending on model needs.

Timelag column

Stored as object strings like '+1:06:27'.

Needs conversion to numeric seconds or minutes for modeling.

Rolling metrics

No NaNs, but max values capped at 999, likely placeholders for missing or uninitialized data.

Numeric & categorical features

All pps_* and rdr_* columns numeric ✅

Other categorical columns like Team, Stage_Name, Race_Name contain missing values.

In [10]:
# =========================
# 0️⃣ Imports
# =========================
import pandas as pd
import numpy as np
import os
import re

# =========================
# 1️⃣ Load existing processed data
# =========================
data_path = "../phase2/processed_data/race_data_processed.csv"
df = pd.read_csv(data_path, parse_dates=['Date','birthdate'])
print(f"Original shape: {df.shape}")

# =========================
# 2️⃣ Fix age inconsistencies
# =========================
# Clip age_at_race to reasonable range
df['age_at_race'] = df['age_at_race'].clip(lower=15, upper=50)

# For rows with missing birthdate, fill age_at_race with Age
df.loc[df['birthdate'].isna(), 'age_at_race'] = df.loc[df['birthdate'].isna(), 'Age']

# =========================
# 3️⃣ Remove duplicate entries
# =========================
df = df.drop_duplicates(subset=['Rider','Race_ID','Stage_Number'])
print(f"Shape after removing duplicates: {df.shape}")

# =========================
# 4️⃣ Convert Timelag and Time to numeric seconds (safe version)
# =========================
def parse_timedelta_safe(t):
    if pd.isna(t):
        return 0
    t = str(t).strip().replace('+','').replace('-','').replace('–','')
    numbers = re.findall(r'\d+', t)
    if not numbers:
        return 0
    if len(numbers) == 3:
        return int(numbers[0])*3600 + int(numbers[1])*60 + int(numbers[2])
    elif len(numbers) == 2:
        return int(numbers[0])*60 + int(numbers[1])
    elif len(numbers) == 1:
        return int(numbers[0])
    return 0

df['Timelag_sec'] = df['Timelag'].apply(parse_timedelta_safe)
df['Time_sec'] = df['Time'].apply(parse_timedelta_safe)

print("Sample converted Timelag/Time:")
print(df[['Timelag','Timelag_sec','Time','Time_sec']].head(10))

# =========================
# 5️⃣ Optional: Fill missing categorical values
# =========================
df['Team'] = df['Team'].fillna('Unknown')
df['Stage_Name'] = df['Stage_Name'].fillna('Unknown')
df['country'] = df['country'].fillna('Unknown')

# =========================
# 6️⃣ Save new processed CSV
# =========================
output_dir = "../phase2/processed_data"
os.makedirs(output_dir, exist_ok=True)
output_path = os.path.join(output_dir, "race_data_processed_v3.csv")

df.to_csv(output_path, index=False)
print(f"New processed data saved to: {output_path}")
print(f"New shape: {df.shape}")


Original shape: (219423, 54)
Shape after removing duplicates: (151746, 54)
Sample converted Timelag/Time:
  Timelag  Timelag_sec   Time  Time_sec
0   +0:10           10   0:00         0
1   +3:05          185   3:55       235
2   +3:13          193   0:00         0
3  +17:21         1041  14:16       856
4  +22:50         1370   5:31       331
5  +25:53         1553   3:03       183
6   +3:21          201   2:29       149
7   +5:31          331   2:10       130
8  +15:27          927   9:56       596
9  +33:18         1998  17:53      1073
New processed data saved to: ../../Joel/processed_data\race_data_processed_v3.csv
New shape: (151746, 56)
